state-chain/chains/src/btc/smart_contract_encoding.rs

```diff
@@ -0,0 +1,95 @@
+use crate::ForeignChainAddress;
+use cf_primitives::{Asset, AssetAmount};
+use codec::{Decode, Encode};
+
+use super::{BitcoinOp, BitcoinScript};
+
+// The maximum length of data that can be encoded in a nulldata utxo
+const MAX_NULLDATA_UTXO_LENGTH: usize = 80;
+
+#[derive(Encode, Decode, Clone)]
+pub struct UtxoEncodedData {
+	pub output_asset: Asset,
+	pub output_address: ForeignChainAddress,
```

Comment:
ForeignChainAddress is an internal type, not really for external use. It also duplicates information already held in the asset field (the asset implies the chain so we don't need to encode this again in the address). 

How about we move the address to the end and assume that any trailing bytes are part of the address? 

```rust
pub struct UtxoEncodedData {
	pub parameters: SharedCfParameters,
	pub output_asset: Asset,
	// Assumed to be the raw bytes, 20 bytes for evm, 32 for dot/sol
	// btc should be rejected
	pub output_address: Vec<u8>,
}
```

Alternatively we decode the bytes based on the output asset:

```rust
pub struct UtxoEncodedData {
	pub parameters: SharedCfParameters,
	pub output_asset: Asset,
	pub output_address: EncodedAddress,
}

impl Decode for UtxoEncodedData {
	// Custom impl that uses try_into_encoded_address
	// See implementation in evm::vault.rs
}
```

Reply:
Forgot to mention that this has also been updated @dandanlen